### PR TITLE
Simulation

### PR DESCRIPTION
**Pull request by schuemie**
Adding a simulator that can generate data in CDM format. There are currently two simulation tasks: generating data for pretraining, and generating data for prediction (fine-tuning a pre-trained model and testing the fine-tuned model).

The simulator uses hidden states, which can lead to observed concepts. Once a concept is observed, observing other concepts becomes more likely. Any set of observed concepts on a day is combined into a visit.

The main Python files are 

- simulation/simulation.py, containing the main simulation code, and
- cdm_data.py, containing classes for growing CDM data and writing those to Parquet (as expected by the CDM processor)

All options for the simulator are controlled via the simulation.ini file.

### CODE REVIEW

File: simulation/simulation.py [hunk 1]
```diff
@@ -0,0 +1,400 @@
+"""
+Simulate CDM data for testing purposes. Persons are simulated to have hidden disease states. Some states are fixes at
```
Comment (egillax) [inline simulation/simulation.py:2]:
```suggestion
Simulate CDM data for testing purposes. Persons are simulated to have hidden disease states. Some states are fixed at
```

File: simulation/simulation.py [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+        self.visit_multiplier = visit_multiplier
+        self.days_to_simulate = days_to_simulate
+
+
+class SimulationTask:
+    partition_count: int
+
+
+class PreTrainingTask(SimulationTask):
+    person_count: int
```
Comment (egillax) [inline simulation/simulation.py:66]: Is there a reason you are using a mix of [data class](https://docs.python.org/3/library/dataclasses.html) features and init methods.  Like here you have `person_count: int` under the class definition and `self.person_count = person_count` in the `init` method. The latter is redundant and will be automatically generated from the former. 

In fact you could remove the init methods from `SimulationSettings` and `PreTrainingTask` if you add the variables under the class definition.

I think it would be good to be consistent in using either a data class or init methods where the init methods don't contain any logic, only assignment to class variables.

I haven't used data classes much myself so I could be missing something.

File: simulation/simulation.py [hunk 3]
```diff
@@ -0,0 +1,400 @@
...
+                                  month_of_birth=month_of_birth,
+                                  day_of_birth=day_of_birth,
+                                  gender_concept_id=gender_concept_id)
+        visit_occurrence_id = person_id * 100000
+        for t in range(self._settings.days_to_simulate):
+            if self._log_verbosity == OBSESSIVE:
+                logging.debug("Person: %s, Day: %s, State: %s", person_id, t, state)
+            state_interaction_matrix = np.outer(state, state)
+            state_flip_probabilities = np.zeros(shape=self._settings.dynamic_state_count)
+            # Role dice to enter a dynamic state:
```
Comment (egillax) [inline simulation/simulation.py:211]:
```suggestion
            # Roll dice to enter a dynamic state:
```

File: simulation/simulation.py [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+            for i in range(self._settings.dynamic_state_count):
+                if state[i] == 1:
+                    continue
+                value = -6 + (state_interaction_matrix * self._dynamic_state_entry_coefs[i]).sum()
+                state_flip_probabilities[i] = logistic(value)
+            # Role dice to exit a dynamic state:
```
Comment (egillax) [inline simulation/simulation.py:217]:
```suggestion
            # Roll dice to exit a dynamic state:
```

File: simulation/simulation.py [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+        for t in range(self._settings.days_to_simulate):
+            if self._log_verbosity == OBSESSIVE:
+                logging.debug("Person: %s, Day: %s, State: %s", person_id, t, state)
+            state_interaction_matrix = np.outer(state, state)
+            state_flip_probabilities = np.zeros(shape=self._settings.dynamic_state_count)
+            # Role dice to enter a dynamic state:
+            for i in range(self._settings.dynamic_state_count):
+                if state[i] == 1:
+                    continue
+                value = -6 + (state_interaction_matrix * self._dynamic_state_entry_coefs[i]).sum()
```
Comment (egillax) [inline simulation/simulation.py:215]: Why is this -6 there?

File: simulation/simulation.py (hunk 2 above)
Comment (schuemie) [reply simulation/simulation.py:66]: Thanks. I'm pretty new to Python, so wasn't aware of the dataclasses module. I'll make sure the data classes only have data.

**Comment by egillax** [comment]
@schuemie I think this looks good. I'm not really following the monte carlo stuff though. I fixed a few typos and had one question about your use of data classes and another one about magic numbers in the simulation code. 

I'm going to try this today and see how it works.

File: setup.py [hunk 6]
```diff
@@ -1,6 +1,6 @@
 from distutils.core import setup
 
-setup(name='GeneralPretrainedModelTools',
+setup(name='Apollo',
       version='1.0',
-      packages=["cdm_processing", "utils"],
+      packages=["cdm_processing", "utils", "simulation", "data_generation"],
```
Comment (egillax) [inline setup.py:5]: There's a mismatch here. The folder with data_generation code is called "data_generating". This gives an error when trying to install the package. I would be in favor of using "data_generation", so renaming the folder.

File: simulation/simulation.py [hunk 7]
```diff
@@ -0,0 +1,400 @@
...
+        self._initial_state_probabilities = np.concatenate((
+            np.random.beta(size=settings.dynamic_state_count, a=1, b=4) / settings.dynamic_state_count,
+            np.random.uniform(size=settings.fixed_state_count + 2)
+        ))
+        # Note: really only need half of these matrices to be filled, because interaction matrix will be symmetrical
+        # over diagonal, but it's easier to just fill them completely:
+        self._dynamic_state_entry_coefs = []
+        for i in range(settings.dynamic_state_count):
+            matrix = np.zeros(self._state_count * self._state_count)
+            non_zero_count = round(len(matrix) / self._state_count)
```
Comment (ChaoPang) [inline simulation/simulation.py:158]: Nitpick: Isn't `non_zero_count` equal to `self._state_count`?
```suggestion
            non_zero_count = self._state_count
```

File: simulation/simulation.py [hunk 8]
```diff
@@ -0,0 +1,400 @@
...
+        else:
+            raise ValueError("Unknown task type: %s" % type(self._task))
+        logging.debug("Finished partition %s of %s", partition_i, self._task.partition_count)
+
+    def _simulate_person_set(self,
+                             partition_i: int,
+                             persons_per_partition: int,
+                             output_folder: str):
+        if isinstance(self._task, PredictionTask):
+            self._cdm_data = cdm_data.CdmDataWithLabels()
```
Comment (ChaoPang) [inline simulation/simulation.py:336]: A quick question about how `multiprocessing.Pool` works with the class method `_simulate_partition`. After creating multiple processes, each process will call `_simulate_person_set`, and re-instantiate `self._cdm_data`. I wonder if multiple processes will compete for the class member `self._cdm_data` and cause data inconsistency. Is the `self` object actually getting cloned to each individual process?

File: simulation/cdm_data.py [hunk 9]
```diff
@@ -0,0 +1,206 @@
...
+class CdmDataWithLabels(CdmData):
+
+    def __init__(self):
+        super().__init__()
+        self._concept_id_to_labels: defaultdict[int, _Labels] = defaultdict(_Labels)
+
+    def add_label(self, person_id: int, concept_id: int, label: bool):
+        labels = self._concept_id_to_labels[concept_id]
+        labels.person_id.append(person_id)
+        labels.label.append(label)
```
Comment (ChaoPang) [inline simulation/cdm_data.py:187]: It looks like we store a `person_id` and `label` pair for every `concept_id`, e.g. there are 100 folders generated containing the same `person_id` and `label` pairs after running the simulation, could you elaborate a bit more on how this data will be used for downstream prediction?

**Review by ChaoPang — CHANGES_REQUESTED** [review]
Thanks for putting this together, I learned a lot from going through your code. Similarly, I can't say I understand exactly how the simulation model works, I would like to get your slides on this if possible. Nevertheless, I left a couple of comments and suggestions below related to Numpy

File: simulation/simulation.py (hunk 7 above)
Comment (schuemie) [reply simulation/simulation.py:158]: Yes, I see that is true now. But I was trying to figure out what proportion of the matrix should be empty. At first I took a fixed proportion (e.g. size / 100), now I tried to make it more flexible by dividing by the state count. May still choose to change that, so I'll keep it as it is.

File: simulation/simulation.py (hunk 8 above)
Comment (schuemie) [reply simulation/simulation.py:336]: Yes, Pool creates a clone, so everything set before Pool will be available to all threads. Everything set in a thread stays in the thread, unless you explicitly create shared memory objects.

File: simulation/simulation.py [hunk 10]
```diff
@@ -0,0 +1,400 @@
...
+            prediction_labels = None
+            index_day = 0
+            is_prediction = False
+        start_date = np.datetime64("2010-01-01") + np.random.randint(0, 365 * 10)
+        state = np.random.binomial(n=1, p=self._initial_state_probabilities)
+        self._cdm_data.add_observation_period(person_id=person_id,
+                                              observation_period_id=person_id,
+                                              observation_period_start_date=start_date,
+                                              observation_period_end_date=start_date + self._settings.days_to_simulate)
+        year_of_birth, month_of_birth, day_of_birth = _simulate_date_of_birth(start_date, state[-2])
```
[the remark this answers is not in the document]
Comment (schuemie) [reply simulation/simulation.py:198]: Will do!

File: simulation/simulation.py [hunk 11]
```diff
@@ -0,0 +1,400 @@
...
+        gender_concept_id = 8507 if state[-1] == 0 else 8532
```
[the remark this answers is not in the document]
Comment (schuemie) [reply simulation/simulation.py:199]: Will do!

File: simulation/simulation.py [hunk 12]
```diff
@@ -0,0 +1,400 @@
...
+                else:
+                    concept_ids = self._concept_ids[observed_concepts != 0]
+                    for concept_id in concept_ids:
+                        self._cdm_data.add_condition_occurrence(person_id=person_id,
+                                                                condition_start_date=start_date + t,
+                                                                condition_concept_id=concept_id,
+                                                                visit_occurrence_id=visit_occurrence_id)
+                    # If admission concept is serious, make the visit an emergency room visit, otherwise make it an
+                    # outpatient visit:
+                    if np.in1d(self._concept_ids[admission_concepts != 0], self._serious_concepts).any():
```
[the remark this answers is not in the document]
Comment (schuemie) [reply simulation/simulation.py:248]: This is not entirely correct, since the `concept_ids` variable also includes non-admission concepts, whereas only the admission concepts determine the visit type. 

However, I will change the code so it uses `admission_concepts`, which I'll rename to `admission_concept_idx` just to avoid confusion between concept IDs and (boolean) indexes of said concept IDs.

File: setup.py (hunk 6 above)
Comment (schuemie) [reply setup.py:5]: Agreed. For consistency, I'll rename `simulation` to `simulating`, and `simulation.py` to `simulator.py` as well.

File: simulation/simulation.py [hunk 13]
```diff
@@ -0,0 +1,400 @@
...
+                    continue
+                state_flip_probabilities[i] = self._dynamic_state_exit_probabilities[i]
+            state[:self._settings.dynamic_state_count] = np.logical_xor(state[:self._settings.dynamic_state_count],
+                                                                        np.random.binomial(n=1,
+                                                                                           p=state_flip_probabilities))
+            # Role dice to observe a concept:
+            concept_probabilities = np.zeros(shape=self._settings.concept_count)
+            for i in range(self._settings.concept_count):
+                value = -8 + (state_interaction_matrix * self._concept_emmision_coefs[i]).sum()
+                concept_probabilities[i] = logistic(value)
```
[the remark this answers is not in the document]
Comment (schuemie) [reply simulation/simulation.py:229]: Nice! I'll incorporate, but will move the intercept (-8) to the front

File: simulation/simulation.py [hunk 14]
```diff
@@ -0,0 +1,400 @@
...
+            for i in range(self._settings.dynamic_state_count):
+                if state[i] == 1:
+                    continue
+                value = -6 + (state_interaction_matrix * self._dynamic_state_entry_coefs[i]).sum()
+                state_flip_probabilities[i] = logistic(value)
+            # Role dice to exit a dynamic state:
+            for i in range(self._settings.dynamic_state_count):
+                if state[i] == 0:
+                    continue
+                state_flip_probabilities[i] = self._dynamic_state_exit_probabilities[i]
```
[the remark this answers is not in the document]
Comment (schuemie) [reply simulation/simulation.py]: Nice! I'll copy this (I now have too many local changes to merge your change directly)

File: simulation/cdm_data.py [hunk 15]
```diff
@@ -0,0 +1,206 @@
...
+from collections import defaultdict
+from typing import Type
+import logging
+
+import numpy as np
+import pyarrow as pa
+import pyarrow.parquet as pq
+
+
+class DynamicArray:
```
[the remark this answers is not in the document]
Comment (schuemie) [reply simulation/cdm_data.py:16]: Yes, I was surprised there's nothing like it is base Python. I'll add some unit tests.

File: simulation/cdm_data.py (hunk 9 above)
Comment (schuemie) [reply simulation/cdm_data.py:187]: Each folder is a partitioned Parquet table corresponding to a concept ID. in each table the label for a person ID will be True if the concept ID was observed in the prediction window, and False if it was not. So we can use this to train and validate a prediction model (and one way to train a prediction model is to fine-tune a pre-trained model)

**Comment by schuemie** [comment]
Thanks @egillax  and @ChaoPang for your feedback and changes! I think I now addressed all. I'll merge this into main.
